src/i2s.rs

```diff
@@ -0,0 +1,1835 @@
+            }
+        }
+    }
+
+    /// I2S pulse density modulation (PDM) downsampling mode.
+    #[cfg(esp_idf_soc_i2s_supports_pdm_rx)]
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum PdmDownsample {
+        /// Downsample 8 samples.
+        Samples8,
+
+        /// Downsample 16 samples.
+        Samples16,
+
+        /// Maximum downsample rate.
+        Max,
+    }
+
+    #[cfg(esp_idf_soc_i2s_supports_pdm_rx)]
+    impl Default for PdmDownsample {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Samples8
+        }
+    }
+
+    #[cfg(esp_idf_soc_i2s_supports_pdm_rx)]
+    impl PdmDownsample {
+        /// Convert to the ESP-IDF SDK `i2s_pdm_downsample_t` representation.
+        #[inline(always)]
+        #[allow(unused)] // TODO: remove when PDM is implemented.
+        pub(super) fn as_sdk(&self) -> esp_idf_sys::i2s_pdm_dsr_t {
+            match self {
+                Self::Samples8 => 0,
+                Self::Samples16 => 1,
+                Self::Max => 2,
+            }
+        }
+    }
+
+    /// Pulse density modulation (PDM) transmit signal scaling mode.
+    #[cfg(esp_idf_soc_i2s_supports_pdm_tx)]
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum PdmSignalScale {
+        /// Divide the PDM signal by 2.
+        Div2,
+
+        /// No scaling.
+        None,
+
+        /// Multiply the PDM signal by 2.
+        Mul2,
+
+        /// Multiply the PDM signal by 4.
+        Mul4,
+    }
+
+    #[cfg(esp_idf_soc_i2s_supports_pdm_tx)]
+    impl Default for PdmSignalScale {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::None
+        }
+    }
+
+    #[cfg(esp_idf_soc_i2s_supports_pdm_tx)]
+    impl PdmSignalScale {
+        /// Convert to the ESP-IDF SDK `i2s_pdm_signal_scale_t` representation.
+        #[inline(always)]
+        #[allow(unused)] // TODO: remove when PDM is implemented.
+        pub(crate) fn as_sdk(&self) -> esp_idf_sys::i2s_pdm_sig_scale_t {
+            match self {
+                Self::Div2 => 0,
+                Self::None => 1,
+                Self::Mul2 => 2,
+                Self::Mul4 => 3,
+            }
+        }
+    }
+
+    /// I2S channel operating role
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum Role {
+        /// Controller (master)
+        Controller,
+
+        /// Target (slave)
+        Target,
+    }
+
+    impl Default for Role {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Controller
+        }
+    }
+
+    impl Role {
+        /// Convert to the ESP-IDF SDK `i2s_role_t` representation.
+        #[inline(always)]
+        pub(super) fn as_sdk(&self) -> i2s_role_t {
+            match self {
+                Self::Controller => I2S_ROLE_CONTROLLER,
+                Self::Target => I2S_ROLE_TARGET,
+            }
+        }
+    }
+
+    /// Standard mode channel configuration.
+    pub struct StdChanConfig<Bclk, Din, Dout, Mclk, Ws>
```

Comment:
Let's try to remove all pins from this struct. Let's instead just have them in the `new` constructor of the driver. Once we have support for other modes, we can have a separate constructor for each mode.

Reply:
So, before we get too far down this path, keep in mind the particular use case I'm hoping to eventually use this for, which is parallel I2S, which is *not* currently supported by ESP-IDF.

[This is the code that sets it up](https://github.com/Xinyuan-LilyGO/LilyGo-EPD47/blob/master/src/i2s_data_bus.c#L211-L235) for the display I'm working with. `gpio_setup_out()` sets the pin up as an output and calls the matrix code to route the pin:
```
    PIN_FUNC_SELECT(GPIO_PIN_MUX_REG[gpio], PIN_FUNC_GPIO);
    gpio_set_direction(gpio, GPIO_MODE_DEF_OUTPUT);
    gpio_matrix_out(gpio, sig, invert, false);
```

As [this separate C driver notes](https://github.com/TobleMiner/esp_i2s_parallel), ESP32's I2S parallel mode is a bit hairy with regard to data widths and clock speeds.

As I write this, I *think* I'm arguing myself toward exactly what you're proposing, @ivmarkov, with separate drivers for the different modes, but would like to confirm.

Reply:
Gone in the latest rewrite.